docs/modules/ROOT/pages/upgrades.adoc

```diff
@@ -0,0 +1,113 @@
+
+=== Usage
+
+Integrating this module into a contract first requires supporting some access control logic, in order to
+avoid unauthorized upgrades. The {ownable} module is used for this example.
+
+[,javascript]
+----
+#[starknet::contract]
+mod UpgradeableContract {
+    use openzeppelin::access::ownable::Ownable;
+    use openzeppelin::upgrades::Upgradeable;
+    use starknet::ClassHash;
+    use starknet::ContractAddress;
+
+    #[storage]
+    struct Storage {}
+
+    #[constructor]
+    fn constructor(self: @ContractState, owner: ContractAddress) {
+        let mut unsafe_state = Ownable::unsafe_new_contract_state();
+        Ownable::InternalImpl::initializer(ref unsafe_state, owner);
+    }
+
+    #[external(v0)]
+    fn upgrade(ref self: ContractState, new_class_hash: ClassHash) {
+        // This function can only be called by the owner
+        let ownable_state = Ownable::unsafe_new_contract_state();
+        Ownable::InternalImpl::assert_only_owner(@ownable_state);
+
+        // Replace the class hash upgrading the contract
+        let mut upgradeable_state = Upgradeable::unsafe_new_contract_state();
+        Upgradeable::InternalImpl::upgrade(ref upgradeable_state, new_class_hash);
+    }
+
+    (...)
+}
+----
+
+=== Interface
+
+Implementing this interface to support upgradeability, supports interoperability by for example allowing DAOs
+and plugins to connect to contracts and upgrade them.
+
+[,javascript]
+----
+trait IUpgradeable {
+    fn upgrade(new_class_hash: ClassHash);
+}
+----
+
+== Proxies and Starknet
+
+A proxy contract is a contract that delegates function calls to another contract.
+This type of pattern decouples state and logic.
+Proxy contracts store the state and redirect function calls to an implementation contract that handles the logic,
+usually catching these calls with a fallback mechanism.
+This allows for different patterns such as upgrades, where implementation contracts can change but the proxy contract (and thus the state) does not;
+as well as deploying multiple proxy instances pointing to the same implementation.
+This can be useful to deploy many contracts with identical logic but unique initialization data.
+
+In the case of contract upgrades, it is achieved by simply changing the proxy's reference to the class hash of the declared implementation.
+This allows developers to add features, update logic, and fix bugs without touching the state or the contract address to interact with the application.
+
+In Starknet, implementing a proxy pattern has an important limitation: there is no fallback mechanism to be used
+for redirecting every potential function call to the implementation contract. This means a generic proxy contract
+can't be implemented. Instead, a proxy contract must be implemented with each specific function that is going to be redirected.
+This can still be useful for example with upgrading the logic of some functions.
```

Comment:
i don't think i understand the example. i get you can implement `proxy::foo` such that gets delegated to `impl::foo`, but what's the pattern?

Reply:
The pattern is that you can update the impl class hash in the proxy, effectively upgrading the logic of the function foo, without changing the contract address, by using proxies.

Reply:
You can implement beacon proxies or anything proxy-related similar to Ethereum, with just one restriction: function names need to be known from the very beginning. I think this is still very useful. You can update 20 ERC20Votes contracts by changing the impl for updating the delegate_by_sig method to use a different hash generation mechanism, as one example, but could be many more.

Reply:
> The pattern is that you can update the impl class hash in the proxy, effectively upgrading the logic of the function foo, without changing the contract address, by using proxies.

right but you can do that by simply changing the proxy's own class. What you describe is technically feasible, but i believe we need to think through how would this be implemented and used. Is the value really there?

> You can update 20 ERC20Votes contracts by changing the beacon's implementation (with replace_class_hash) for updating the delegate_by_sig method to use a different hash generation mechanism, as one example, but could be many more.

are you sure about this? i'm not sure you can delegate calls to another contract, just another class.